ext/drbbdup/drbbdup.h

```diff
@@ -0,0 +1,334 @@
+ * drbbdup_analyse_bb_t function.
+ *
+ * @param pre_analysis_data preliminary analysis data.
+ * @param encoding the case encoding that is associated with the analysis data.
+ * @param analysis_data the analysis data to destroy.
+ *
+ * \note The user should not destroy pre_analysis_data.
+ */
+typedef void (*drbbdup_destroy_analysis_t)(void *drcontext, void *user_data,
+                                           uintptr_t encoding, void *pre_analysis_data,
+                                           void *analysis_data);
+
+/**
+ * Inserts code responsible for encoding the current runtime
+ * case. The function should store the resulting pointer-sized encoding to the memory
+ * destination operand obtained via drbbdup_get_encoding_opnd(). If the user has
+ * implemented the encoder via a clean call, drbbdup_set_encoding() should be
+ * used instead.
+ *
+ * @param pre_analysis_data the preliminary analysis data.
+ */
+typedef void (*drbbdup_insert_encode_t)(void *drcontext, instrlist_t *bb, instr_t *where,
+                                        void *user_data, void *pre_analysis_data);
+
+/**
+ * A user-defined call-back function that is invoked to instrument an instruction \p
+ * instr. The inserted code must be placed at \p where.
+ *
+ * Instrumentation must be driven according to the passed case encoding \p encoding.
+ *
+ * @param instr the instr to be instrumented.
+ * @param where the location where to insert instrumentation.
+ * @param encoding the encoding of the current case considered.
+ * @param pre_analysis_data preliminary analysis data.
+ * @param analysis_data analysis data for the particular case.
+ */
+typedef void (*drbbdup_instrument_instr_t)(void *drcontext, instrlist_t *bb,
+                                           instr_t *instr, instr_t *where,
+                                           uintptr_t encoding, void *user_data,
+                                           void *pre_analysis_data, void *analysis_data);
+
+/***************************************************************************
+ * INIT
+ */
+
+/**
+ * Specifies the options when initialising drbbdup. \p set_up_bb_dups, \p insert_encode
+ * and \p instrument_instr cannot be NULL, while \p dup_limit must be greater than zero.
+ */
+typedef struct {
+    /** Set this to the size of this structure. */
+    size_t struct_size;
+    /** A user-defined call-back function that sets up how to duplicate a basic block.
+     * Cannot be NULL.
+     */
+    drbbdup_set_up_bb_dups_t set_up_bb_dups;
+    /**
+     * A user-defined call-back function that inserts code to encode the runtime case.
+     * The resulting encoding is used by the dispatcher to direct control to the
+     * appropriate basic block. Cannot be NULL.
+     */
+    drbbdup_insert_encode_t insert_encode;
+    /**
+     * A user-defined call-back function that conducts a preliminary analysis on a basic
+     * block.
+     */
+    drbbdup_pre_analyse_bb_t pre_analyse_bb;
```

Comment:
I wonder if you could describe why these analysis control points need to be brokered by drbbdup instead of just being up to the client?

Reply:
To clarify, I don't mean in a comment here in the code, I mean to explain to me (since there's no .dox) since I'm not understanding the full picture yet.

Reply:
Are you asking about pre_analyze_bb_t functions in particular? This analysis is done per fragment, whilst drbbdup_analyse_bb_t functions are with respect to different case encodings. You would conduct a pre analysis if you are doing some computation (usually expensive) that is relevant to all cases. In other words, rather than repeating the same analysis for each basic block copy, you do it once within a pre_analyse_bb function. When you say that the client can broker this type analysis, how would you go about it? Remember, at this stage, the app2app phase has passed and the fragment consists of duplications. drbbdup takes care of extracting a single bb and passing it to the user via a drbbdup_pre_analyse_bb_t call-back.

Reply:
Ah ok, that makes it clearer.

WDYT of s/pre_analyze/analyze_orig/ or sthg: since it's not about being before the drmgr analysis phase, it's about analyzing the original single-copy code.  And s/analyze_bb/analyze_one_case/ or sthg.  Those name would have made it easier for me to grasp their purpose.

Reply:
How about analyze_case instead of analyze_one_case?

Reply:
Done.

Reply:
"analyze_case" SGTM